compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -1339,12 +1330,30 @@ object Parsers {
           WhileDo(cond, body)
         }
       case DO =>
-        atSpan(in.skipToken()) {
+        in.errorOrMigrationWarning(
+          i"""`do <body> while <cond>' is no longer supported,
+             |use `while {<body> ; <cond>} do ()' instead.
+             |The statement can be rewritten automatically under -language:Scala2 -migration -rewrite.
+           """)
+        val start = in.skipToken()
+        atSpan(start) {
           val body = expr()
           if (isStatSep) in.nextToken()
+          val whileStart = in.offset
           accept(WHILE)
           val cond = expr()
-          DoWhile(body, cond)
+          if (ctx.settings.migration.value) {
+            patch(source, Span(start, start + 2), "while ({")
+            patch(source, Span(whileStart, whileStart + 5), ";")
+            cond match {
+              case Parens(_) =>
+                patch(source, Span(cond.span.start, cond.span.start + 1), "")
+                patch(source, Span(cond.span.end - 1, cond.span.end), "")
+              case _ =>
+            }
+            patch(source, cond.span.endPos, "}) ()")
+          }
+          WhileDo(Block(body :: Nil, cond), Literal(Constant(())))
```

Comment:
This is incorrect, as both `body` and `block` could define symbols with the same name. If they are put together in a single `Block`, the scopes will clash.

Reply:
We will need a test for this.

Reply:
Actually, `body` and `cond` are already blocks if they have definitions. I added some tests in #7002. Did I miss some other definitions that could be represented differently?